sui/wormhole/sources/governance/governance_message.move

```diff
@@ -0,0 +1,471 @@
+
+    const E_OLD_GUARDIAN_SET_GOVERNANCE: u64 = 0;
+    const E_INVALID_GOVERNANCE_CHAIN: u64 = 1;
+    const E_INVALID_GOVERNANCE_EMITTER: u64 = 2;
+    const E_INVALID_GOVERNANCE_MODULE: u64 = 4;
+    const E_INVALID_GOVERNANCE_ACTION: u64 = 5;
+    const E_GOVERNANCE_TARGET_CHAIN_NONZERO: u64 = 6;
+    const E_GOVERNANCE_TARGET_CHAIN_NOT_SUI: u64 = 7;
+
+    struct GovernanceMessage {
+        module_name: Bytes32,
+        action: u8,
+        chain: u16,
+        payload: vector<u8>,
+        vaa_hash: Bytes32
+    }
+
+    public fun module_name(self: &GovernanceMessage): Bytes32 {
+        self.module_name
+    }
+
+    public fun action(self: &GovernanceMessage): u8 {
+        self.action
+    }
+
+    public fun is_global_action(self: &GovernanceMessage): bool {
+        self.chain == 0
+    }
+
+    public fun is_local_action(self: &GovernanceMessage): bool {
+        self.chain == chain_id()
+    }
+
+    public fun vaa_hash(self: &GovernanceMessage): Bytes32 {
+        self.vaa_hash
+    }
+
+    #[test_only]
+    public fun payload(self: &GovernanceMessage): vector<u8> {
+        self.payload
+    }
+
+    public fun take_payload(msg: GovernanceMessage): vector<u8> {
+        let GovernanceMessage {
+            module_name: _,
+            action: _,
+            chain: _,
+            vaa_hash: _,
+            payload
+        } = msg;
+
+        payload
+    }
+
+    public fun parse_and_verify_vaa(
+        wormhole_state: &mut State,
+        vaa_buf: vector<u8>,
+        ctx: &TxContext
+    ): GovernanceMessage {
+        let parsed =
+            vaa::parse_and_verify(
+                wormhole_state,
+                vaa_buf,
+                ctx
+            );
+
+        // This VAA must have originated from the governance emitter.
+        assert_governance_emitter(wormhole_state, &parsed);
+
+        let vaa_hash = vaa::hash(&parsed);
+
+        let cur = cursor::new(vaa::take_payload(parsed));
+
+        let module_name = bytes32::take(&mut cur);
+        let action = bytes::take_u8(&mut cur);
+        let chain = bytes::take_u16_be(&mut cur);
+        let payload = cursor::rest(cur);
+
+        GovernanceMessage { module_name, action, chain, payload, vaa_hash }
+    }
+
+    public fun take_global_action(
+        msg: GovernanceMessage,
+        expected_module_name: Bytes32,
+        expected_action: u8
+    ): vector<u8> {
+        assert_module_and_action(&msg, expected_module_name, expected_action);
+
+        // New guardian sets are applied to all Wormhole contracts.
+        assert!(is_global_action(&msg), E_GOVERNANCE_TARGET_CHAIN_NONZERO);
+
+        take_payload(msg)
+    }
+
+    public fun take_local_action(
+        msg: GovernanceMessage,
+        expected_module_name: Bytes32,
+        expected_action: u8
+    ): vector<u8> {
+        assert_module_and_action(&msg, expected_module_name, expected_action);
+
+        // New guardian sets are applied to all Wormhole contracts.
```

Comment:
Should this comment be the same as the one in line #96? For local action, isn't it just applied to this chain?

`New guardian sets are applied to all Wormhole contracts`

Reply:
Maybe this comment should be moved to a more suitable location, like in `update_guardian_set`

Reply:
Didn't mean to live this. Fixed in [46ecb65](https://github.com/wormhole-foundation/wormhole/pull/2494/commits/46ecb659639869cddd1bf525097f2b3f676b459a)